c/tests/check_grounding_space.c

```diff
@@ -97,12 +103,291 @@ START_TEST (test_replace)
 }
 END_TEST
 
+START_TEST (test_abstract_space_grounding)
+{
+    space_t* space = space_new_grounding_space();
+    space_add(space, expr(atom_sym("+"), atom_var("a"), atom_sym("B"), 0));
+    atom_t* query = expr(atom_sym("+"), atom_sym("A"), atom_var("b"), 0);
+
+    struct output_t result = { "", 0 };
+    space_query(space, query, query_callback, &result);
+    ck_assert_str_eq(result.str, "b: B, ");
+
+    space_free(space);
+}
+END_TEST
+
+void str_to_buf(const char *str, void *context) {
+    strncpy(context, str, BUF_SIZE);
+};
+
+// QUESTION FOR VITALY: What's the simplest way to test the observer mechanism?  Obviously I could
+// add my own observer, but I assume there is a better in-situ test that is possible?
```

Comment:
Not sure I understand the question right. I would add an observer, modify the space by adding, removing and replacing something and check that observers was called.

Reply:
I had some confusion about the intended use for the space observer mechanism. Specifically, I wasn’t sure if the idea was:

1. the API's user can implement their own observers and observe spaces they did not implement, or 
2. the API’s user can implement their own spaces, and those spaces must provide notifications to all observers.

From reading your comments, it sounds like both are possible and both must be supported.  The original code supported 2, but not yet 1.  I have just committed support for 1.

Reply:
For reference, those changes were the purpose of 9699e67

Reply:
> From reading your comments, it sounds like both are possible and both must be supported. The original code supported 2, but not yet 1. I have just committed support for 1.

Ideally both should be supported, but main priority is to support the case 2. I was hoping supporting 1 would be natural, but you have found it is not (at least with current automatic deregistration API). As I said in https://github.com/trueagi-io/hyperon-experimental/pull/284#discussion_r1184570787 I would be even glad to move to the option you suggested when observer ownership is given to the space. It should simplify implementing 1.

Reply:
Let me know what you think of 4d1ec5d.  Unless there is a reason for C space implementations to keep a hold of all observers, I think this simplifies the C API in a good way.